File: src/main/java/frc/robot/Vision.java
package frc.robot;

import edu.wpi.first.cameraserver.CameraServer;
import edu.wpi.first.cscore.CvSink;
import edu.wpi.first.cscore.CvSource;
import edu.wpi.first.cscore.UsbCamera;
import edu.wpi.first.cscore.VideoCamera;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

import org.opencv.core.Mat;
import org.opencv.core.Point;
import org.opencv.core.Scalar;
import org.opencv.core.Size;
import org.opencv.core.MatOfPoint;
import org.opencv.core.Rect;
import org.opencv.imgproc.Imgproc;
import org.opencv.core.Core;

import java.util.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

public class Vision {

    // static singleton vision class instance
    private static Vision instance = null;
    private static Constants C = new Constants();
    // Public Variables
    public int imgWidth;
    public int imgHeight;
    public Rect biggestRed;
    public Rect biggestYellow;
    public Rect biggestGreen;
    public Rect[] biggestColors = { biggestRed, biggestYellow, biggestGreen };
    // Private Variables
    private Thread visionThread;
    private RectAreaComparator rectSorter;
    // private Size gb = new Size(7, 7); // gaussian blur
    // Private Final Variables
    // private final Rect EMPTY_RECT = new Rect(0, 0, 0, 0);
    private Mat BLACK_IMAGE;

    /*
     * Constructors
     */
    private Vision(int imageWidth, int imageHeight) {
        this.imgWidth = imageWidth;
        this.imgHeight = imageHeight;
        this.rectSorter = new RectAreaComparator();
        this.init();
    }

    private Vision() {
        this(320, 240);
    }

    /*
     * Accessor to Vision
     * - uses singleton design pattern
     */
    public static Vision getInstance() {
        if (instance == null) {
            instance = new Vision();
        }

        return instance;
    }

    public static Vision getInstance(int a, int b) {
        if (instance == null) {
            instance = new Vision(a, b);
        }

        return instance;
    }

    /*
     * Rect Comparator Class
     * - compares Rect from smallest to largest area
     */
    class RectAreaComparator implements Comparator<Rect> {

        @Override
        public int compare(Rect r1, Rect r2) {
            return (int) (r1.area() - r2.area());
        }

    }

    /*
     * Color Detector Class
     * - detects specified color
     * - assumes gaussian blur has been applies
     * 
     * 1. thresholds on color
     * 2. finds contours
     * 3. finds bounding boxes on contours
     * 4. updates to list of detected rectangles & biggest rectangle color
     */

    class ColorDetector {
        private Scalar color;
        private Scalar upperLimit;
        private Scalar lowerLimit;
        private Mat mask;
        private Mat tempImage;
        private List<MatOfPoint> contours;
        private Mat hierarchy;
        private Rect biggestRect;
        private List<Rect> detectedRects;
        private Rect br;
        private Rect temp;

        public ColorDetector(Scalar c, Scalar ll, Scalar ul) {
            color = c;
            upperLimit = ul;
            lowerLimit = ll;

            mask = new Mat();
            hierarchy = new Mat();
            contours = new ArrayList<MatOfPoint>();
            detectedRects = new ArrayList<Rect>();
        }

        private void reset() {
            BLACK_IMAGE.copyTo(mask);
            BLACK_IMAGE.copyTo(hierarchy);
            contours.clear();
        }

        private Rect findRects(Mat image) {

            reset();

            Core.inRange(image, this.lowerLimit, this.upperLimit, this.mask);

            Imgproc.findContours(this.mask, this.contours, this.hierarchy, Imgproc.RETR_EXTERNAL,
                    Imgproc.CHAIN_APPROX_SIMPLE);

            // BLACK_IMAGE.copyTo(tempImage);

            if (this.contours.size() == 0) {
                temp = C.EMPTY_RECT;
            } else {
                this.temp = C.EMPTY_RECT;
                this.detectedRects.clear();
                for (int i = 0; i < Math.min(this.contours.size(), C.MAX_OBJECT_DETECT); i++) {
                    // Imgproc.drawContours(sourceMat, contours, i, redColor, 2, Imgproc.LINE_8,
                    // hierarchy, 0, new Point());
                    br = Imgproc.boundingRect(this.contours.get(i));
                    this.detectedRects.add(br);
                    if (br.area() > temp.area()) {
                        this.temp = br;
                    }
                    if (C.VERBOSE) {
                        Imgproc.rectangle(image, br.tl(), br.br(), color, 1);
                    }
                }

                // sort detected rectangles from biggest to smallest area
                Collections.sort(this.detectedRects, rectSorter);
                Collections.reverse(this.detectedRects);
            }
            this.biggestRect = this.temp;

            return this.biggestRect;
        }
    }

    /*
     * Calculate center of rectangle
     */
    public Point rectCenter(Rect r) {
        return new Point(r.tl().x + (r.width / 2.0), r.tl().y + (r.height / 2.0));
    }

    /*
     * Convert Point to array [x, y]
     * 
     */
    public double[] point2array(Point p) {
        double[] arr = { p.x, p.y };

        return arr;
    }

    /*
     * Initializes Computer Vision
     */
    private void init() {

        // add USB camera, create server for SmartDashboard
        UsbCamera usbCamera = CameraServer.startAutomaticCapture("Main Camera", 0);
        usbCamera.setResolution(imgWidth, imgHeight);
        usbCamera.setExposureManual(5);
        usbCamera.setWhiteBalanceManual(3300);
        usbCamera.setBrightness(60); // default 50
        usbCamera.getProperty("focus_auto").set(0);
        usbCamera.getProperty("saturation").set(100);
        

        CvSink cvSink = CameraServer.getVideo(); // grab images from camera
        CvSource outputStream = CameraServer.putVideo("Processed Image", imgWidth, imgHeight);

        // Init variables
        Mat sourceMat = new Mat();
        Mat outputMat =new Mat();
        Mat tempMat = new Mat();
        BLACK_IMAGE = Mat.zeros(imgHeight, imgWidth, 16);
        BLACK_IMAGE.copyTo(sourceMat);
        BLACK_IMAGE.copyTo(outputMat);
        BLACK_IMAGE.copyTo(tempMat);
        Point centerTop = new Point(160,0);
        Point centerBottom = new Point(160,240);
        Point centerTopLeft = new Point(160 - C.centerThresh, 0);
        Point centerBottomRight = new Point(160 + C.centerThresh, 240);

        ColorDetector redDetector = new ColorDetector(C.RED_COLOR, C.redLower, C.redUpper);
        ColorDetector yellowDetector = new ColorDetector(C.YELLOW_COLOR, C.yellowLower, C.yellowUpper);
        ColorDetector greenDetector = new ColorDetector(C.GREEN_COLOR, C.greenLower, C.greenUpper);

        List<ColorDetector> detectors = new ArrayList<ColorDetector>();
        detectors.add(redDetector);
        detectors.add(yellowDetector);
        detectors.add(greenDetector);

        int numDetectors = detectors.size();
        
        visionThread = new Thread(() -> {

            while (true) { /// TODO: change condition later
                // System.out.println(usbCamera.getPath());
                long startTime = System.currentTimeMillis();

                if (cvSink.grabFrame(sourceMat) != 0) {

                    sourceMat.copyTo(tempMat);

                    // gaussian blur
                    Imgproc.GaussianBlur(sourceMat, sourceMat, C.gb, 0, 0);


                    BLACK_IMAGE.copyTo(outputMat);

                    // color detection
                    for (int i = 0; i < numDetectors; i++) {
                        // for(int i = 0; i < 1; i++){
                        ColorDetector cd = detectors.get(i);
                        biggestColors[i] = cd.findRects(sourceMat);
                        // System.out.println("found red");
                        outputMat.setTo(cd.color, cd.mask);
                        if (!C.VERBOSE) {
                            Imgproc.rectangle(outputMat, biggestColors[i].tl(), biggestColors[i].br(), cd.color, 1);
                            Imgproc.circle(outputMat, rectCenter(biggestColors[i]), 3, C.BLUE_COLOR, -1);
                            Imgproc.rectangle(tempMat, biggestColors[i].tl(), biggestColors[i].br(), cd.color, 1);
                            Imgproc.circle(tempMat, rectCenter(biggestColors[i]), 3, C.BLUE_COLOR, -1);
                        }
                    }

                    biggestRed = biggestColors[0];
                    biggestYellow = biggestColors[1];
                    biggestGreen = biggestColors[2];

                    Imgproc.line(tempMat, centerTop, centerBottom, C.WHITE_COLOR, 1);
                    Imgproc.rectangle(tempMat, centerTopLeft, centerBottomRight, C.WHITE_COLOR, 1);

                    // output results
                    outputStream.putFrame(tempMat); // put processed image to smartdashboard
                    long endTime = System.currentTimeMillis();

                    System.out.println("Total execution time: " + (endTime - startTime));

                    SmartDashboard.putString("Biggest Red Center", rectCenter(biggestRed).toString());
                    SmartDashboard.putString("Biggest Yellow Center", rectCenter(biggestYellow).toString());
                    SmartDashboard.putString("Biggest Green Center", rectCenter(biggestGreen).toString());
                }

            }

        });

        visionThread.setPriority(10); // highest priority
    }

    public void startVision() {
        visionThread.start(); // start vision thread
    }

}

File: src/main/java/frc/robot/vision.py
import json
import time
import sys
import cv2
import numpy as np

from cscore import CameraServer, UsbCamera
import ntcore

configFile = "/boot/frc.json"
WIDTH = 640
HEIGHT = 480
GB = (11, 11)

redlow = (0, 70, 110)
redhigh = (8, 255, 255)

redlow2 = (160, 190, 120)
redhigh2 = (179, 255, 255)

yellowlow = (22, 70, 110)
yellowhigh = (35, 255, 255)

biggestAreaRed = 0.0
biggestXRed = 0.0
biggestYRed = 0.0

biggestAreaYellow = 0.0
biggestXYellow = 0.0
biggestYYellow = 0.0

DEBUG = True
THRESH = 15

def getBiggestColor(mask, output_img):
    contours, hierarchy = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    # cv2.drawContours(output_img, contours, -1, (0,255,0), 3)

    biggestArea = 0.0
    biggestX = 0.0
    biggestY = 0.0
    biggestHTemp = 0.0
    biggestWTemp = 0.0
    tl = ()
    br = ()
    cent_coord = ()
    # print(len(contours))
    for c in contours:
        x,y,w,h = cv2.boundingRect(c)
        # cv2.rectangle(output_img,(x,y),(x+w,y+h),(0,255,0),2)
        a = w*h
        centerx = x + (w/2)
        centery = y + (h/2)
        if a > biggestArea:
            biggestArea = a
            biggestX = centerx
            biggestY = centery
            biggestHTemp = h
            biggestWTemp = w
            tl = (x, y)
            br = (x + w, y + h)
            cent_coord = (int(x + w/2), int(y + h/2))
    
    if (len(tl) != 0 and  len(br) != 0 and len(cent_coord) != 0):
        cv2.rectangle(output_img, tl, br,(0,0,255),2)
        cv2.circle(output_img, cent_coord, 5, (255,0,0), -1)
    # print(cent_coord)
    # print(br)

    cv2.line(output_img, (WIDTH//2, 0), (WIDTH//2, HEIGHT), (255, 255, 255), 2)
    cv2.line(output_img, (WIDTH//2 + THRESH, 0), (WIDTH//2 + THRESH, HEIGHT), (255, 255, 255), 1)
    cv2.line(output_img, (WIDTH//2 - THRESH, 0), (WIDTH//2 - THRESH, HEIGHT), (255, 255, 255), 1)

    return biggestArea, biggestX, biggestY, output_img

def debugVision(input_stream, output_stream, img):

    blur = np.zeros((WIDTH, HEIGHT, 3), dtype=np.uint8)
    hsv = np.zeros((WIDTH, HEIGHT, 3), dtype=np.uint8)
    redmask1 = np.zeros((WIDTH, HEIGHT, 1), dtype=np.uint8)
    redmask2 = np.zeros((WIDTH, HEIGHT, 1), dtype=np.uint8)
    redmask = np.zeros((WIDTH, HEIGHT, 1), dtype=np.uint8)
    yellowmask = np.zeros((WIDTH, HEIGHT, 1), dtype=np.uint8)

    biggestAreaRedTemp = 0.0
    biggestXRedTemp = 0.0
    biggestYRedTemp = 0.0

    biggestAreaYellowTemp = 0.0
    biggestXYellowTemp = 0.0
    biggestYYellowTemp = 0.0

    while True:

        # grab image from camera
        result, img = input_stream.grabFrame(img)
        output_img = np.copy(img)

        if(result != 0):

            start_time = time.time()

            # print('image shape, min, max', img.shape, img.min(axis=(0,1)), img.max(axis=(0,1)))

            blur = cv2.GaussianBlur(img, GB, 0) # apply gaussian blur

            hsv = cv2.cvtColor(blur, cv2.COLOR_BGR2HSV) # convert to HSV colorspace

            # red threshold
            redmask1 = cv2.inRange(hsv, redlow, redhigh)
            redmask2 = cv2.inRange(hsv, redlow2, redhigh2)
            redmask = cv2.bitwise_or(redmask1, redmask2)
            # print('redlow', redlow)
            # print('redhi', redhigh)
            # print('mask', redmask.min(), redmask.max(), redmask.sum())

            # yellow threshold
            yellowmask = cv2.inRange(hsv, yellowlow, yellowhigh)

            biggestAreaRedTemp, biggestXRedTemp, biggestYRedTemp, output_img = getBiggestColor(redmask, output_img)
            biggestAreaYellowTemp, biggestXYellowTemp, biggestYYellowTemp, output_img = getBiggestColor(yellowmask, output_img)


            
            output_stream.putFrame(output_img)
            biggestAreaRed.set(biggestAreaRedTemp)
            biggestXRed.set(biggestXRedTemp)
            biggestYRed.set(biggestYRedTemp)

            biggestAreaYellow.set(biggestAreaYellowTemp)
            biggestXYellow.set(biggestXYellowTemp)
            biggestYYellow.set(biggestYYellowTemp)

            end_time = time.time()

            print(end_time - start_time)


if __name__ == "__main__":
    global biggestAreaRed, biggestAreaYellow, biggestXRed, biggestXYellow, biggestYRed, biggestYYellow
    # print(cv2.__version__)
    with open(configFile) as f:
        config = json.load(f)

    team = config["team"]
    camera_config = config['cameras'][0]

    print("Starting camera '{}' on {}".format(camera_config["name"], camera_config["path"]))

    nt = ntcore.NetworkTableInstance.getDefault()
    # nt.setSeverTeam(6593)
    nt.startClient4("rasp pi client")
    nt.setServerTeam(6593)
    nt.startDSClient()
    # nt.setServer("host", ntcore.NetworkTableInstance.kDefaultPort4)

    # camera.getProperty("name").set(val)
    camera = CameraServer.startAutomaticCapture()
    camera.setResolution(WIDTH, HEIGHT)
    camera.setBrightness(60)
    camera.setWhiteBalanceManual(3300)
    camera.setExposureManual(2)
    camera.getProperty("contrast").set(50)
    camera.getProperty("saturation").set(80)
    camera.getProperty("focus_auto").set(0)
    camera.getProperty("gain").set(50)

    input_stream = CameraServer.getVideo()
    output_stream = CameraServer.putVideo("Processed RaspPi", WIDTH, HEIGHT)

    vision_nt = nt.getTable("Vision")
    biggestAreaRed = vision_nt.getDoubleTopic("redA").publish()
    biggestXRed = vision_nt.getDoubleTopic("redX").publish()
    biggestYRed = vision_nt.getDoubleTopic("redY").publish()

    biggestAreaYellow = vision_nt.getDoubleTopic("yellowA").publish()
    biggestXYellow = vision_nt.getDoubleTopic("yellowX").publish()
    biggestYYellow = vision_nt.getDoubleTopic("yellowY").publish()

    img = np.zeros((WIDTH, HEIGHT, 3), dtype=np.uint8)

    time.sleep(0.5)

    debugVision(input_stream, output_stream, img)